src/lib/message/L2ToL1Message.ts

```diff
@@ -301,54 +300,44 @@ export class L2ToL1MessageReader extends L2ToL1Message {
 
     const l2Network = await getL2Network(l2Provider)
 
-    const rollup = new Contract(
+    const rollup = RollupUserLogic__factory.connect(
       l2Network.ethBridge.rollup,
-      [
-        'function latestConfirmed() public view returns (uint64)',
-        'event NodeConfirmed(uint64 indexed nodeNum, bytes32 blockHash, bytes32 sendRoot)',
-      ],
-      l1Provider
+      this.l1Provider
     )
 
     // CHRIS: TODO: could confirm in between these calls
-    const latestConfirmedNode = await rollup['latestConfirmed']()
-    const currentBlock = await l1Provider.getBlockNumber()
+    const latestConfirmedNodeNum = await rollup.callStatic.latestConfirmed()
+    const latestConfirmedNode = await rollup.getNode(latestConfirmedNodeNum)
 
     // now get the block hash and sendroot for that node
-    const logs = await l1Provider.getLogs({
-      address: rollup.address,
-      fromBlock: Math.max(
-        // CHRIS: TODO: either a constant or think about removing this -it's 4 weeks
-        currentBlock - Math.floor((4 * 7 * 24 * 60 * 60) / 14),
-        0
-      ),
-      toBlock: 'latest',
-      topics: rollup.interface.encodeFilterTopics(
-        rollup.interface.getEvent('NodeConfirmed'),
-        [latestConfirmedNode]
-      ),
-    })
-
-    if (logs.length !== 1) throw new Error('missing logs')
-
-    const parsedLog = rollup.interface.parseLog(logs[0]).args as unknown as {
-      nodeNum: BigNumber
-      sendRoot: string
-      blockHash: string
-    }
+    const eventFetcher = new EventFetcher(l1Provider)
+    const logs = await eventFetcher.getEvents(
+      rollup.address,
+      RollupUserLogic__factory,
+      t => t.filters.NodeCreated(latestConfirmedNodeNum),
+      {
+        fromBlock: latestConfirmedNode.createdAtBlock.toNumber(),
+        toBlock: latestConfirmedNode.createdAtBlock.toNumber(),
+      }
+    )
+
+    if (logs.length !== 1) throw new ArbTsError('No NodeConfirmed events found')
+
+    const parsedLog = parseNodeCreatedAssertion(logs[0])
 
     const l2Block = await (
       l2Provider! as ethers.providers.JsonRpcProvider
-    ).send('eth_getBlockByHash', [parsedLog.blockHash, false])
-    if (l2Block['sendRoot'] !== parsedLog.sendRoot) {
-      console.log(l2Block['sendRoot'], parsedLog.sendRoot)
-      throw new Error('send roots')
+    ).send('eth_getBlockByHash', [parsedLog.afterState.blockHash, false])
+    if (l2Block['sendRoot'] !== parsedLog.afterState.sendRoot) {
+      // CHRIS: TODO: handle this case
+      console.log(l2Block['sendRoot'], parsedLog.afterState.sendRoot)
+      throw new ArbTsError("L2 block send root doesn't match parsed log")
     }
 
     const sendRootSize = BigNumber.from(l2Block['sendCount'])
     if (sendRootSize.gt(this.event.position)) {
       this.sendRootSize = sendRootSize
-      this.sendRootHash = parsedLog.sendRoot
+      this.sendRootHash = parsedLog.afterState.sendRoot
```

Comment:
Not really in the scope of this PR, but any not to have updateSendRoot just be getSendRoot (that returns the two values instead of storing them as class variables?)